### PR TITLE
[v4.0] feat: preserve shebang in entry module for CJS and ESM outputs

### PR DESCRIPTION
**Pull request by TrickyPi**
This PR contains:

- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [x] yes (_breaking changes will not be merged unless absolutely necessary_)
- [ ] no

List any relevant issue numbers:

### CODE REVIEW

File: src/utils/renderChunks.ts [hunk 1]
```diff
@@ -111,6 +111,10 @@ async function transformChunk(
 ) {
 	let map: SourceMap | null = null;
 	const sourcemapChain: DecodedSourceMapOrMissing[] = [];
+	const entryModule = usedModules.find(module => module.info.isEntry);
+	if (entryModule && entryModule.shebang !== undefined && ['cjs', 'es'].includes(options.format)) {
+		magicString.prepend(`#!${entryModule.shebang}\n`);
+	}
```
Comment (lukastaegert) [inline src/utils/renderChunks.ts:117]: Why do it here and not in `Chunk.render` where everything else is rendered?
More specifically, it feels to me like something I would search for in the "finalizers", i.e. `es.ts` and `cis.ts`. You could add the shebang as another property to the `FinaliserOptions` and only render it for those two formats.
Another point to consider: Especially when using manual chunks, a chunk might have several entry points. Those do not matter. The only module that matters is the `facadeModule`, i.e. the module this chunk pretends to be. See also the `isEntryFacade` property.

**Review by lukastaegert — COMMENTED** [review]
The general approach with the Module property looks very good to me, I would slightly move and adjust the rendering logic, though.

File: src/utils/renderChunks.ts (hunk 1 above)
Comment (TrickyPi) [reply src/utils/renderChunks.ts]: Ah, I see, `Chunk.render` is a more appropriate place.

Comment (TrickyPi) [reply src/utils/renderChunks.ts]: > You could add the shebang as another property to the `FinaliserOptions` and only render it for those two formats.

For rendering `shebang` before `banner`, I prepend `shebang` after `banner` in `Chunk.render`.

Comment (lukastaegert) [reply src/utils/renderChunks.ts]: Ah you are right, I missed that banner is attached outside the finalizers. Good work then!

**Review by lukastaegert — APPROVED** [review]

**Comment by zaripych** [comment]
cc @TrickyPi 

What would be the recommended way to remove the shebang from the output?

For context - I use `tsx` to run my TypeScript files in dev mode and have a plugin that adds `#!/bin/env node` shebang after transpilation. The way I do that is by adding a `banner` to the output.

Rollup 4 now breaks this workflow.

**Comment by lukastaegert** [comment]
The [`renderChunk`](https://rollupjs.org/plugin-development/#renderchunk) hook allows you to edit generated chunks.

**Pull request merged**